packages/platforms/react-native/lib/entropy-source.ts

```diff
@@ -0,0 +1,63 @@
+import { type Spec as NativeBugsnagPerformance } from './NativeBugsnagPerformance'
+
+export interface EntropySource {
+  nextByte: () => number
+}
+
+const POOL_SIZE = 1024
+
+// fill the pool with random values
+export function fill (pool: Uint8Array) {
+  for (let i = 0; i < pool.length; i++) {
+    pool[i] = (Math.random() * 255) | 0
+  }
+}
+
+// shuffle the existing values in the pool
+export function shuffle (pool: Uint8Array) {
+  for (let i = 0; i < POOL_SIZE; i++) {
+    const otherIndex = (Math.random() * (POOL_SIZE - 1)) | 0
+    const otherValue = pool[otherIndex]
+    pool[otherIndex] = pool[i]
+    pool[i] = otherValue
+  }
+}
+
+const createEntropySource = (NativeBugsnag: NativeBugsnagPerformance | null): EntropySource => {
+  const pool = new Uint8Array(POOL_SIZE)
+
+  // initialise the pool synchronously
+  const randomValues = NativeBugsnag && NativeBugsnag.requestEntropy()
+  randomValues && randomValues.length > 0 ? pool.set(randomValues) : fill(pool)
+
+  // if the native module is not available, we will regenerate the pool synchronously in JS,
+  // otherwise we will regenerate asynchronously from the native module
+  let regenerate = () => { fill(pool) }
+  if (NativeBugsnag) {
+    regenerate = async () => {
+      const randomValues = await NativeBugsnag.requestEntropyAsync()
+      randomValues.length > 0 ? pool.set(randomValues) : fill(pool)
+    }
+  }
+
+  let nextIndex = 0
+  const nextByte = () => {
+    const value = pool[nextIndex]
+    nextIndex++
+
+    // when the pool is exhausted, trigger an async operation to regenerate.
+    // We also shuffle the current pool so that we don't have to wait for
```

Comment:
If the pool's exhausted why does shuffling it help?

Reply:
The call to the native layer is expensive, so shuffling the pool means we don't need to await the async call and can continue to generate IDs in the meantime

Reply:
I must be misunderstanding.  Wouldn't that result in using an id you've already used?

Reply:
no - it's not a pool of IDs, it's a sequence of random bytes that is used to generate the IDs. Each ID uses either 8 or 16 bytes  (see `id-generator.ts`)

so by shuffling we get a different sequence of bytes while we wait for a new pool

Reply:
if it's an asynchronous call, would it be better to rehydrate the pool after a certain number _near_ the limit has been reached? then we wouldn't need to reshuffle the values we've already got, we can just grab the next allocation, and any subsequent call should have a rehydrated source.

Reply:
the issue with that is you don't know how long the call will take or how many IDs you will need to generate before it completes